addons/account/models/account_move_line.py

```diff
@@ -1089,6 +1089,17 @@ def _check_off_balance(self):
                 if line.reconciled:
                     raise UserError(_('Lines from "Off-Balance Sheet" accounts cannot be reconciled'))
 
+    @api.constrains('account_id', 'display_type')
+    def _check_payable_receivable(self):
+        for line in self:
+            account_type = line.account_id.account_type
+            if line.move_id.is_sale_document(include_receipts=True):
+                if (line.display_type == 'payment_term') ^ (account_type == 'asset_receivable'):
+                    raise UserError(_("You can only set an account having the receivable type on payment terms lines for customer invoice."))
+            if line.move_id.is_purchase_document(include_receipts=True):
+                if (line.display_type == 'payment_term') ^ (account_type == 'liability_payable'):
+                    raise UserError(_("You can only set an account having the payable type on payment terms lines for vendor bill."))
```

Comment:
The message might need a little refresher, but I'm not sure what would be the clearest way to formulate this

Reply:
```suggestion
                    raise UserError(_("Any journal item on a Payable Account must have a due date and vice versa."))
```

and of course : "Any journal item on a Receivable Account must have a due date and vice versa." on Sales Documents.

Reply:
It's really about the type of the account on the lines that refer to the payment terms (al the lines with payable/receivable accounts), not about the date

Reply:
Yes, but functionally speaking, the user has no way to know a line refers to payment term... Functionally speaking, it's a term because of a due date.